### PR TITLE
WIP: Wavelet surrogates

### PR DESCRIPTION
**Pull request by kahaaga**
Implements the wavelet based surrogates from [Keylock (2006)](https://journals.aps.org/pre/abstract/10.1103/PhysRevE.73.036707).

## TODO:

- [ ] It is not clear to me how to perform the last iteration step.
- [ ] The circular shifting step is waaaay to slow. This can be done more efficiently.

### CODE REVIEW

**Comment by kahaaga** [comment]
The method seems to work well to preserve the autocorrelation structure of the original data, even without the last iteration step from Keylock (2006).

```julia 
using TimeseriesSurrogates, Random
Random.seed!(1234);
n = 400
σ = 20
x = cumsum(randn(n)) .+ 
    [20*sin(2π/10*i) for i = 1:n] .+ 
    [20*cos(2π/50*i) for i = 1:n] .+
    [50*sin(2π/2*i + π) for i = 1:n] .+ 
    σ .* rand(n).^2;

s = surrogate(x, WLS(AAFT()));
surroplot(x, s)
```
Screenshot 2020-05-16 07 24 19

I state this explicitly in the documentation that this step is not performed. Maybe that can do for now. Otherwise, I have to think about this a bit more. Hm.

File: src/methods/wavelet_based.jl [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+preserves nonstationarity. 
+
+To deal with nonstationary signals, Keylock (2006) recommends using a 
+wavelet with a high number of vanishing moments. Thus, default here is 
+a Daubechies wavelet with 16 vanishing moments.
+
+[^Keylock2006]: C.J. Keylock (2006). "Constrained surrogate time series with preservation of the mean and variance structure". Phys. Rev. E. 73: 036707. doi:10.1103/PhysRevE.73.036707.
+"""
+struct WLS <: Surrogate
+    wt::Wavelets.WT.OrthoWaveletClass
```
Comment (Datseris) [inline src/methods/wavelet_based.jl:31]: Isn't this an abstract type? You have to make it a type parameter instead.

```WLS{W <: Wavelets.WT.OrthoWaveletClass
wt::W
```

If not we have type instability which leads to significant performance drop.

Comment (kahaaga) [reply src/methods/wavelet_based.jl]: You are right. Fixed.

File: src/methods/wavelet_based.jl [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+"""
+struct WLS{WT <: Wavelets.WT.OrthoWaveletClass} <: Surrogate
+    wt::WT
+    surromethod::Surrogate # should preserve values of the original series
+end
+
+# Keylock (2006) recommends a wavelet with high number of vanishing moments,
+# so use Daubechies wavelets with 16 vanishing moments as default.
+WLS() = WLS(Wavelets.WT.Daubechies{16}(), AAFT())
+WLS(surromethod::Surrogate) = WLS(Wavelets.WT.Daubechies{16}(), surromethod)
```
Comment (Datseris) [inline src/methods/wavelet_based.jl:38]: I am not sure this order of arguments is the best... If you just define a single 
```
WLS(surr = AAFT(), wave = Daubechies{16}())
```
you can use the same method in the docstring, thus showing the defaults as well. I was bit confused reading the documentation page and seeing a `<:Surrogate` passed as the first argument everywhere.

**Comment by kahaaga** [comment]
@Datseris The wavelet surrogates are supposed to preserve the local mean and variance of the original signal. Here, I am not consistently getting those properties preserved, so I am not convinced my implementation is correct yet. I'm also not able to reproduce the examples in the original paper with the last step of the algorithm missing. 

Therefore, I think this should be postponed to a later release.  We can do 1.0 without the wavelet-based surrogates for now.

**Comment by Datseris** [comment]
Okay, sounds good

**Comment by kahaaga** [comment]
@Datseris I take it back! I just realized I wasn't actually _performing_ the error minimization which is required to fit the local properties of the signal :grimacing: Fixed it by matching detail coefficients at each dyadic scale by maximizing a simple correlation, and now it works as perfectly, also for non-stationary signals (see below)! 

This should definitely be in 1.0! I will just add a few notes to the documentation, then this can be merged.

```julia
using TimeseriesSurrogates, Random
Random.seed!(5040)
n = 500
σ = 30
x = cumsum(randn(n)) .+ 
    [20*sin(2π/30*i) for i = 1:n] .+ 
    [20*cos(2π/90*i) for i = 1:n] .+
    [50*sin(2π/2*i + π) for i = 1:n] .+ 
    σ .* rand(n).^2 .+ 
    [0.5*t for t = 1:n];

method = WLS(IAAFT(), true)
s = surrogate(x, method);
p = surroplot(x, s)
```
Screenshot 2020-05-23 17 32 10

**Pull request merged**